# EDA on the Dataset

In [1]:
#Change directory and load catalog and pipeline from Kedro
%cd "D:\University\CST4090 - Thesis\Propensity Model\thesis-propensity-model"
%reload_kedro

D:\University\CST4090 - Thesis\Propensity Model\thesis-propensity-model


                    INFO     No path argument was provided. Using: D:\University\CST4090 -            ]8;id=694097;file://C:\Users\Nijel\Anaconda3\envs\thesis-env\lib\site-packages\kedro\extras\extensions\ipython.py\ipython.py]8;;\:]8;id=181477;file://C:\Users\Nijel\Anaconda3\envs\thesis-env\lib\site-packages\kedro\extras\extensions\ipython.py#55\55]8;;\
                             Thesis\Propensity Model\thesis-propensity-model                                       

[07/25/22 22:59:25] INFO     Kedro project thesis_propensity_model                                    ]8;id=153410;file://C:\Users\Nijel\Anaconda3\envs\thesis-env\lib\site-packages\kedro\extras\extensions\ipython.py\ipython.py]8;;\:]8;id=842753;file://C:\Users\Nijel\Anaconda3\envs\thesis-env\lib\site-packages\kedro\extras\extensions\ipython.py#78\78]8;;\

                    INFO     Defined global variable 'context', 'session', 'catalog' and 'pipelines'  ]8;id=16251;file://C:\Users\Nijel\Anaconda3\envs\thesis-env\lib\site-packages\kedro\extras\extensions\ipython.py\ipython.py]8;;\:]8;id=634668;file://C:\Users\Nijel\Anaconda3\envs\thesis-env\lib\site-packages\kedro\extras\extensions\ipython.py#79\79]8;;\

                    INFO     Registered line magic 'reload_kedro_mlflow'                              ]8;id=575820;file://C:\Users\Nijel\Anaconda3\envs\thesis-env\lib\site-packages\kedro\extras\extensions\ipython.py\ipython.py]8;;\:]8;id=32789;file://C:\Users\Nijel\Anaconda3\envs\thesis-env\lib\site-packages\kedro\extras\extensions\ipython.py#85\85]8;;\

In [2]:
#Load the Raw Dataset
bank_raw=catalog.load("bank_raw")

[07/25/22 22:59:27] INFO     Loading data from 'bank_raw' (CSVDataSet)...                       ]8;id=388886;file://C:\Users\Nijel\Anaconda3\envs\thesis-env\lib\site-packages\kedro\io\data_catalog.py\data_catalog.py]8;;\:]8;id=731794;file://C:\Users\Nijel\Anaconda3\envs\thesis-env\lib\site-packages\kedro\io\data_catalog.py#343\343]8;;\

In [3]:
bank_raw.shape

(41188, 21)

In [4]:
bank_raw.columns.to_list()

[
    'age',
    'job',
    'marital',
    'education',
    'default',
    'housing',
    'loan',
    'contact',
    'month',
    'day_of_week',
    'duration',
    'campaign',
    'pdays',
    'previous',
    'poutcome',
    'emp.var.rate',
    'cons.price.idx',
    'cons.conf.idx',
    'euribor3m',
    'nr.employed',
    'y'
]

In [8]:
bank_raw.head(10)

age 
 job 
 marital 
 education 
 default 
 housing 
 loan 
 contact 
 month 
 day_of_week 
 ... 
 campaign 
 pdays 
 previous 
 poutcome 
 emp.var.rate 
 cons.price.idx 
 cons.conf.idx 
 euribor3m 
 nr.employed 
 y 
 
 
 
 
 0 
 56 
 housemaid 
 married 
 basic.4y 
 no 
 no 
 no 
 telephone 
 may 
 mon 
 ... 
 1 
 999 
 0 
 nonexistent 
 1.1 
 93.994 
 -36.4 
 4.857 
 5191.0 
 no 
 
 
 1 
 57 
 services 
 married 
 high.school 
 unknown 
 no 
 no 
 telephone 
 may 
 mon 
 ... 
 1 
 999 
 0 
 nonexistent 
 1.1 
 93.994 
 -36.4 
 4.857 
 5191.0 
 no 
 
 
 2 
 37 
 services 
 married 
 high.school 
 no 
 yes 
 no 
 telephone 
 may 
 mon 
 ... 
 1 
 999 
 0 
 nonexistent 
 1.1 
 93.994 
 -36.4 
 4.857 
 5191.0 
 no 
 
 
 3 
 40 
 admin. 
 married 
 basic.6y 
 no 
 no 
 no 
 telephone 
 may 
 mon 
 ... 
 1 
 999 
 0 
 nonexistent 
 1.1 
 93.994 
 -36.4 
 4.857 
 5191.0 
 no 
 
 
 4 
 56 
 services 
 married 
 high.school 
 no 
 no 
 yes 
 telephone 
 may 
 mon 
 ... 
 1 
 999 
 0 
 nonexistent 
 1.1 
 93.994 
 -36.4 
 4.857 
 5191.0 
 no 
 
 
 5 
 45 
 services 
 married 
 basic.9y 
 unknown 
 no 
 no 
 telephone 
 may 
 mon 
 ... 
 1 
 999 
 0 
 nonexistent 
 1.1 
 93.994 
 -36.4 
 4.857 
 5191.0 
 no 
 
 
 6 
 59 
 admin. 
 married 
 professional.course 
 no 
 no 
 no 
 telephone 
 may 
 mon 
 ... 
 1 
 999 
 0 
 nonexistent 
 1.1 
 93.994 
 -36.4 
 4.857 
 5191.0 
 no 
 
 
 7 
 41 
 blue-collar 
 married 
 unknown 
 unknown 
 no 
 no 
 telephone 
 may 
 mon 
 ... 
 1 
 999 
 0 
 nonexistent 
 1.1 
 93.994 
 -36.4 
 4.857 
 5191.0 
 no 
 
 
 8 
 24 
 technician 
 single 
 professional.course 
 no 
 yes 
 no 
 telephone 
 may 
 mon 
 ... 
 1 
 999 
 0 
 nonexistent 
 1.1 
 93.994 
 -36.4 
 4.857 
 5191.0 
 no 
 
 
 9 
 25 
 services 
 single 
 high.school 
 no 
 yes 
 no 
 telephone 
 may 
 mon 
 ... 
 1 
 999 
 0 
 nonexistent 
 1.1 
 93.994 
 -36.4 
 4.857 
 5191.0 
 no 
 
 
 
 10 rows × 21 columns

In [9]:
bank_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

In [12]:
bank_raw.describe()

age 
 duration 
 campaign 
 pdays 
 previous 
 emp.var.rate 
 cons.price.idx 
 cons.conf.idx 
 euribor3m 
 nr.employed 
 
 
 
 
 count 
 41188.00000 
 41188.000000 
 41188.000000 
 41188.000000 
 41188.000000 
 41188.000000 
 41188.000000 
 41188.000000 
 41188.000000 
 41188.000000 
 
 
 mean 
 40.02406 
 258.285010 
 2.567593 
 962.475454 
 0.172963 
 0.081886 
 93.575664 
 -40.502600 
 3.621291 
 5167.035911 
 
 
 std 
 10.42125 
 259.279249 
 2.770014 
 186.910907 
 0.494901 
 1.570960 
 0.578840 
 4.628198 
 1.734447 
 72.251528 
 
 
 min 
 17.00000 
 0.000000 
 1.000000 
 0.000000 
 0.000000 
 -3.400000 
 92.201000 
 -50.800000 
 0.634000 
 4963.600000 
 
 
 25% 
 32.00000 
 102.000000 
 1.000000 
 999.000000 
 0.000000 
 -1.800000 
 93.075000 
 -42.700000 
 1.344000 
 5099.100000 
 
 
 50% 
 38.00000 
 180.000000 
 2.000000 
 999.000000 
 0.000000 
 1.100000 
 93.749000 
 -41.800000 
 4.857000 
 5191.000000 
 
 
 75% 
 47.00000 
 319.000000 
 3.000000 
 999.000000 
 0.000000 
 1.400000 
 93.994000 
 -36.400000 
 4.961000 
 5228.100000 
 
 
 max 
 98.00000 
 4918.000000 
 56.000000 
 999.000000 
 7.000000 
 1.400000 
 94.767000 
 -26.900000 
 5.045000 
 5228.100000

In [13]:
bank_raw.describe(include=["object"])

job 
 marital 
 education 
 default 
 housing 
 loan 
 contact 
 month 
 day_of_week 
 poutcome 
 y 
 
 
 
 
 count 
 41188 
 41188 
 41188 
 41188 
 41188 
 41188 
 41188 
 41188 
 41188 
 41188 
 41188 
 
 
 unique 
 12 
 4 
 8 
 3 
 3 
 3 
 2 
 10 
 5 
 3 
 2 
 
 
 top 
 admin. 
 married 
 university.degree 
 no 
 yes 
 no 
 cellular 
 may 
 thu 
 nonexistent 
 no 
 
 
 freq 
 10422 
 24928 
 12168 
 32588 
 21576 
 33950 
 26144 
 13769 
 8623 
 35563 
 36548

In [20]:
for col in bank_raw.columns:
        print(col)
        print(bank_raw[col].value_counts(),end="\n")


age
31    1947
32    1846
33    1833
36    1780
35    1759
      ... 
89       2
91       2
94       1
87       1
95       1
Name: age, Length: 78, dtype: int64
job
admin.           10422
blue-collar       9254
technician        6743
services          3969
management        2924
retired           1720
entrepreneur      1456
self-employed     1421
housemaid         1060
unemployed        1014
student            875
unknown            330
Name: job, dtype: int64
marital
married     24928
single      11568
divorced     4612
unknown        80
Name: marital, dtype: int64
education
university.degree      12168
high.school             9515
basic.9y                6045
professional.course     5243
basic.4y                4176
basic.6y                2292
unknown                 1731
illiterate                18
Name: education, dtype: int64
default
no         32588
unknown     8597
yes            3
Name: default, dtype: int64
housing
yes        21576
no         18622
unknown      990
Name: hou

In [16]:
bank_raw.columns

Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp.var.rate', 'cons.price.idx',
       'cons.conf.idx', 'euribor3m', 'nr.employed', 'y'],
      dtype='object')

age - client's age in years (numeric)

job - type of job (categorical: admin., blue-collar, entrepreneur, housemaid, management, retired, self-employed, services, 
student, technician, unemployed, unknown)

marital - marital status (categorical: divorced, married, single, unknown)

education - client's education (categorical: basic.4y, basic.6y, basic.9y, high.school, illiterate, professional.course, university.degree, unknown)

default - has credit in default? (categorical: no, yes, unknown)

housing - has housing loan? (categorical: no, yes, unknown)

loan - has personal loan? (categorical: no, yes, unknown)

contact - contact communication type (categorical: cellular, telephone)

month - last contact month of the year (categorical: jan, feb, mar, ..., nov, dec)

day_of_week - last contact day of the week (categorical: mon, tue, wed, thu, fri)

duration - last contact duration, in seconds (numeric).

campaign - number of contacts performed and for this client during this campaign (numeric, includes the last contact)

pdays - number of days that have passed after the client was last contacted from the previous campaign (numeric; 999 means the 
client has not been previously contacted)

previous - number of contacts performed for this client before this campaign (numeric)

poutcome - outcome of the previous marketing campaign (categorical: failure, nonexistent, success)

emp.var.rate - employment variation rate, quarterly indicator (numeric)

cons.price.idx - consumer price index, monthly indicator (numeric)

cons.conf.idx - consumer confidence index, monthly indicator (numeric)

euribor3m - euribor 3 month rate, daily indicator (numeric)

nr.employed - number of employees, quarterly indicator (numeric)

Output feature (desired target):

y - has the client subscribed a term deposit? (binary: yes,no)

In [6]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

In [7]:
%matplotlib inline
plt.rcParams["figure.figsize"] = (8, 6)

import warnings
warnings.filterwarnings('ignore')

In [22]:
dict_target = {"no": 0, "yes": 1}
bank_raw["y"] = bank_raw["y"].map(dict_target)
bank_raw.head()

age 
 job 
 marital 
 education 
 default 
 housing 
 loan 
 contact 
 month 
 day_of_week 
 ... 
 campaign 
 pdays 
 previous 
 poutcome 
 emp.var.rate 
 cons.price.idx 
 cons.conf.idx 
 euribor3m 
 nr.employed 
 y 
 
 
 
 
 0 
 56 
 housemaid 
 married 
 basic.4y 
 no 
 no 
 no 
 telephone 
 may 
 mon 
 ... 
 1 
 999 
 0 
 nonexistent 
 1.1 
 93.994 
 -36.4 
 4.857 
 5191.0 
 0 
 
 
 1 
 57 
 services 
 married 
 high.school 
 unknown 
 no 
 no 
 telephone 
 may 
 mon 
 ... 
 1 
 999 
 0 
 nonexistent 
 1.1 
 93.994 
 -36.4 
 4.857 
 5191.0 
 0 
 
 
 2 
 37 
 services 
 married 
 high.school 
 no 
 yes 
 no 
 telephone 
 may 
 mon 
 ... 
 1 
 999 
 0 
 nonexistent 
 1.1 
 93.994 
 -36.4 
 4.857 
 5191.0 
 0 
 
 
 3 
 40 
 admin. 
 married 
 basic.6y 
 no 
 no 
 no 
 telephone 
 may 
 mon 
 ... 
 1 
 999 
 0 
 nonexistent 
 1.1 
 93.994 
 -36.4 
 4.857 
 5191.0 
 0 
 
 
 4 
 56 
 services 
 married 
 high.school 
 no 
 no 
 yes 
 telephone 
 may 
 mon 
 ... 
 1 
 999 
 0 
 nonexistent 
 1.1 
 93.994 
 -36.4 
 4.857 
 5191.0 
 0 
 
 
 
 5 rows × 21 columns

In [23]:
print("Share of customers with product =", '{:.1%}'.format(bank_raw["y"].mean()))

Share of customers with product = 11.3%


In [24]:
bank_raw[bank_raw["y"] == 1].mean() 

age                 40.913147
duration           553.191164
campaign             2.051724
pdays              792.035560
previous             0.492672
emp.var.rate        -1.233448
cons.price.idx      93.354386
cons.conf.idx      -39.789784
euribor3m            2.123135
nr.employed       5095.115991
y                    1.000000
dtype: float64

In [25]:
bank_raw[bank_raw["y"] == 0].mean() 

age                 39.911185
duration           220.844807
campaign             2.633085
pdays              984.113878
previous             0.132374
emp.var.rate         0.248875
cons.price.idx      93.603757
cons.conf.idx      -40.593097
euribor3m            3.811491
nr.employed       5176.166600
y                    0.000000
dtype: float64